### PR TITLE
Make model exporting asynchronous

### LINKED ISSUE #18476: `LegacyExporter.Export()` blocks calling thread causing freezes when exporting large archives
Original discussion mentions skin export from settings menu, but this is also the case for the "Export package" menu option in editor.

### Discussed in https://github.com/ppy/osu/discussions/18475

<div type='discussions-op-text'>

<sup>Originally posted by **cdwcgt** May 29, 2022</sup>
On some computers with slow hard drives, `Export selected skin` will make main thread blocked(osu! not response)
We can make it as a task so it will not dragging down the main thread</div>

### PR DESCRIPTION
**Pull request by cdwcgt**
close #18476 

~~At the beginning, it was written according to `LegacyImportManager`.~~
~~so I create a `LegacyExportManager`, but seen unnecessary.~~

~~But doing this prevents getting dependencies again on every component that needs to be used, might be useful?~~

removed

https://user-images.githubusercontent.com/34775378/202493609-59f5a2b0-fe42-4f6b-b3e7-62663bd128cf.mp4

### CODE REVIEW

File: osu.Game/Database/RealmAccess.cs [hunk 1]
```diff
@@ -173,11 +173,6 @@ public RealmAccess(Storage storage, string filename, GameThread? updateThread =
             if (!Filename.EndsWith(realm_extension, StringComparison.Ordinal))
                 Filename += realm_extension;
 
-#if DEBUG
-            if (!DebugUtils.IsNUnitRunning)
-                applyFilenameSchemaSuffix(ref Filename);
-#endif
```
Comment (bdach) [inline osu.Game/Database/RealmAccess.cs:179]: What is the purpose of this change?

File: osu.Game/Database/LegacyExportManager.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                    await new LegacySkinExporter(exportStorage, realmAccess, notification, stream).ExportASync(item);
+                    break;
+
+                case ScoreInfo:
+                    await new LegacyScoreExporter(exportStorage, realmAccess, notification, stream).ExportASync(item);
+                    break;
+
+                case BeatmapSetInfo:
+                    await new LegacyBeatmapExporter(exportStorage, realmAccess, notification, stream).ExportASync(item);
+                    break;
```
Comment (bdach) [inline osu.Game/Database/LegacyExportManager.cs:61]: Not sure I see much point to this class.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Database/RealmAccess.cs (hunk 1 above)
Comment (cdwcgt) [reply osu.Game/Database/RealmAccess.cs:179]: This should be a wrong change. Maybe I accidentally commited it during the test

File: osu.Game/Database/LegacyModelExporter.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        protected readonly Storage ExportStorage;
+
+        protected readonly RealmAccess RealmAccess;
+
+        protected readonly ProgressNotification Notification;
+
+        protected string Filename = null!;
+
+        protected Stream? OutputStream;
```
[the remark this answers is not in the document]
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs:37]: At first, I was set to a private class, but when rewriting the method, I found that these attributes were needed, so I changed to protected.

The rewriting method should be reconsidered..

**Comment by bdach** [comment]
@peppy before this goes too far, curious of your thoughts on how this should be structured. I gave a rough sketch of how I'd see this parceled in the review above but I'm not super sure if I'm not overengineering this. If you're fine with the local resolves peppered around then maybe this is fine as a start.

File: osu.Game/Database/LegacyModelExporter.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            UserFileStorage = storage.GetStorageForDirectory(@"files");
+            RealmAccess = realm;
+
+            notification = new ProgressNotification
+            {
+                State = ProgressNotificationState.Active,
+                Text = "Exporting...",
+                CompletionText = "Export completed"
+            };
+            notification.CancelRequested += () => canCancel;
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:53]: How does this even work? It's checking whether it.. can cancel but not actually performing any cancellation?

File: osu.Game/Database/LegacyModelExporter.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                {
+                    await ExportToStreamAsync(model, stream);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Export te model corresponding to uuid to given stream.
+        /// </summary>
+        /// <param name="uuid">The medel which have <see cref="IHasGuidPrimaryKey"/>.</param>
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:79]: The model to export?

File: osu.Game/Database/LegacyModelExporter.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        /// <param name="stream">The stream to export.</param>
+        /// <returns></returns>
+        public virtual async Task ExportToStreamAsync(TModel uuid, Stream stream)
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:82]:
```suggestion
        public virtual async Task ExportToStreamAsync(TModel model, Stream stream)
```

File: osu.Game/Database/LegacyModelExporter.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            {
+                RealmAccess.Run(r =>
+                {
+                    TModel model = r.Find<TModel>(id);
+                    createZipArchive(model, stream);
+                });
+            }).ContinueWith(OnComplete);
+        }
+
+        private void createZipArchive(TModel model, Stream outputStream)
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Database/LegacyModelExporter.cs:95]: Seems about correct.

File: osu.Game/Online/Leaderboards/LeaderboardScore.cs [hunk 8]
```diff
@@ -66,15 +67,21 @@ public class LeaderboardScore : OsuClickableContainer, IHasContextMenu, IHasCust
...
         private SongSelect songSelect { get; set; }
 
         [Resolved]
         private Storage storage { get; set; }
 
+        [Resolved]
+        private RealmAccess realm { get; set; }
+
+        [Resolved(canBeNull: true)]
+        private INotificationOverlay notifications { get; set; }
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/LeaderboardScore.cs:83]: I tend to agree that making part of the `Manager` classes is best. Going forward, we'll probably have an `ExportLegacy` and an `Export` (for the newer format type, if/when that exists) and it makes sense to keep them both in the same place.

Put another way, this isn't doing anything *so legacy* that it needs to be separate code-side (unlike some of the other stuff which is in the legacy namespace).

File: osu.Game/Database/LegacyModelExporter.cs (hunk 5 above)
Comment (bdach) [reply osu.Game/Database/LegacyModelExporter.cs:79]: FWIW I basically ignored all xmldocs when reviewing this and will probably rewrite them in a final pass. Basically only wanted to ask about structure for now.

File: osu.Game/Database/LegacyModelExporter.cs (hunk 4 above)
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs:53]: I use `notification.CancellationToken.IsCancellationRequested` to confirm whether task is request cancel.

when export zip stream, the task cannot be cancelled, so I use `canCancel` to avoid user cancel the task

**Comment by cdwcgt** [comment]
So should we use DI? Or a manager class or keep.
I make a solution that uses DI, but it feels less than ideal in terms of code quality I think?

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 9]
```diff
@@ -446,6 +453,8 @@ public void UndeleteAll()
         public Task<Live<BeatmapSetInfo>?> ImportAsUpdate(ProgressNotification notification, ImportTask importTask, BeatmapSetInfo original) =>
             beatmapImporter.ImportAsUpdate(notification, importTask, original);
 
+        public void Export(BeatmapSetInfo beatmap) => Task.Run(() => beatmapExporter.ExportAsync(beatmap));
```
Comment (bdach) [inline osu.Game/Beatmaps/BeatmapManager.cs:456]: The threading here is making zero sense. The `ExportAsync()` call (which is *async*) is being wrapped in a `Task.Run()`? And then there is another `Task.Run()` *inside* of `ExportAsync()`? So two TPL threads are now involved? What is going on here?

This should most likely not use `Task.Run()`, and not return `void`. Instead it should return `Task`, and usage sites that don't support async well should use the `.FireAndForget()` extension to get at least a semblance of proper error checking.

File: osu.Game/Online/Leaderboards/LeaderboardScore.cs [hunk 10]
```diff
@@ -96,6 +93,8 @@ private void load(IAPIProvider api, OsuColour colour, ScoreManager scoreManager)
 
             statisticsLabels = GetStatistics(Score).Select(s => new ScoreComponentLabel(s)).ToList();
 
+            this.scoreManager = scoreManager;
```
Comment (bdach) [inline osu.Game/Online/Leaderboards/LeaderboardScore.cs:96]: If you're going to do this then you may as well move `ScoreManager` out from the BDL param to a `[Resolved]` field.

File: osu.Game/Skinning/SkinManager.cs [hunk 11]
```diff
@@ -280,6 +287,8 @@ public Action<IEnumerable<Live<SkinInfo>>> PresentImport
 
         public Task<Live<SkinInfo>> Import(ImportTask task, bool batchImport = false, CancellationToken cancellationToken = default) => skinImporter.Import(task, batchImport, cancellationToken);
 
+        public void ExporCurrenttSkin() => CurrentSkinInfo.Value.PerformRead(s => skinExporter.ExportAsync(s));
```
Comment (bdach) [inline osu.Game/Skinning/SkinManager.cs:290]: Typo in method name.

Also method should arguably not export "current skin", it should export a skin as given by a parameter.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/BeatmapManager.cs (hunk 9 above)
Comment (cdwcgt) [reply osu.Game/Beatmaps/BeatmapManager.cs:456]: I'm dizzy that time, sorry

File: osu.Game/Database/LegacyModelExporter.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Database/LegacyModelExporter.cs]: @cdwcgt still unaddressed

[the remark this answers is not in the document]
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs]: I think its should already use this structure?
image
I don't know what different, please tell me pls

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Database/LegacyModelExporter.cs]: I missed that because there are multiple classes in one file. Please split the files, one class per one file.

[the remark this answers is not in the document]
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs]: ok

File: osu.Game/Database/LegacyArchiveExporter.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+                    {
+                        // Sometimes we cannot find the file(probably deleted by the user), so we handle this and post a error.
+                        if (stream == null)
+                        {
+                            // Only pop up once to prevent spam.
+                            if (!fileMissing)
+                            {
+                                PostNotification?.Invoke(new SimpleErrorNotification
+                                {
+                                    Text = "Some of your files are missing, they will not be included in the archive"
```
Comment (bdach) [inline osu.Game/Database/LegacyArchiveExporter.cs:59]: What on earth? This is shocking. What is this logic even? Why is it here? And when is it ever tolerable to tell a user that "some of their files are missing"?

File: osu.Game/Database/LegacyArchiveExporter.cs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+                }
+            }
+            catch (ObjectDisposedException)
+            {
+                // outputStream may close before writing when request cancel.
+                if (cancellationToken.IsCancellationRequested)
+                    return;
+
+                throw;
+            }
```
Comment (bdach) [inline osu.Game/Database/LegacyArchiveExporter.cs:82]: I don't even know what is happening here.

File: osu.Game/Database/LegacyArchiveExporter.cs (hunk 12 above)
Comment (cdwcgt) [reply osu.Game/Database/LegacyArchiveExporter.cs:59]: This was an accidental case for me, maybe because I imported a skin that used the same file and deleted one, or some other accident, the realm couldn't find the skin file.
If this exception occurs, the entire package will not be exported in old logic, but this is not fatal, so I choose to pop up the prompt and continue

File: osu.Game/Database/LegacyArchiveExporter.cs (hunk 13 above)
Comment (cdwcgt) [reply osu.Game/Database/LegacyArchiveExporter.cs:82]: It looks like if the user cancel
https://github.com/ppy/osu/pull/21308/files#diff-6a0e3884e01dc22403790923a038a2cdda6d02f121b621682c78bc3b12b7c833R100
it will run out of range first and dispose the stream and cause `ObjectDisposedExcption`
Because it can easily happen during the cancellation process, I choose to catch and only return when the user cancels

File: osu.Game/Database/LegacyArchiveExporter.cs (hunk 12 above)
Comment (peppy) [reply osu.Game/Database/LegacyArchiveExporter.cs:59]: I don't yet know how we should handle this kind of thing. I can imagine a scenario where this does happen, and I agree we probably want to present this to the user in some way, but I'm not 100% sure doing it with a notification like this is correct.

Similar cases could come up, for instance, when starting gameplay and files are missing. I'd hope we could standardise the way these are presented across the game. So probably wouldn't be against removing this localised handling for now (or at least replacing with a loud `Logger.Log`).

File: osu.Game/Database/LegacyModelExporter.cs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+            {
+                success = false;
+            }
+            finally
+            {
+                // cleanup if export is failed or canceled.
+                if (!success)
+                {
+                    exportStorage.Delete(filename);
+                }
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:113]: Why not move this outside of the finally and `return` in the `catch`?

File: osu.Game/Database/LegacyModelExporter.cs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+            PostNotification?.Invoke(notification);
+
+            try
+            {
+                using (var stream = exportStorage.CreateFileSafely(filename))
+                {
+                    success = await ExportToStreamAsync(model, stream, notification, cancellationToken ?? notification.CancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:103]: We probably want to look into the best practices of catching `OperationCanceledException`. I'm pretty sure we don't want to catch without a `rethrow`, but we're already doing this (likely incorrectly) in other places.

File: osu.Game/Database/LegacyModelExporter.cs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        /// <param name="stream">The stream to export.</param>
+        /// <param name="notification">The notification will displayed to the user</param>
+        /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
+        /// <returns>Whether the export was successful</returns>
+        public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
+        {
+            ProgressNotification notify = notification ?? new ProgressNotification();
+
+            Guid id = model.ID;
+            return await Task.Run(() =>
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:132]: I'm confused. Why are you using `Task.Run` here? You're in an `async` method. Recommendation would be

```diff
diff --git a/osu.Game/Database/LegacyModelExporter.cs b/osu.Game/Database/LegacyModelExporter.cs
index acb83ce1d5..03f4c06297 100644
--- a/osu.Game/Database/LegacyModelExporter.cs
+++ b/osu.Game/Database/LegacyModelExporter.cs
@@ -124,12 +124,13 @@ public async Task ExportAsync(TModel model, CancellationToken? cancellationToken
         /// <param name="notification">The notification will displayed to the user</param>
         /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
         /// <returns>Whether the export was successful</returns>
-        public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
+        public Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
         {
             ProgressNotification notify = notification ?? new ProgressNotification();
 
             Guid id = model.ID;
-            return await Task.Run(() =>
+
+            return Task.Run(() =>
             {
                 realmAccess.Run(r =>
                 {
@@ -148,7 +149,7 @@ public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, Progres
                 notify.CompletionText = "Export Complete, Click to open the folder";
                 notify.State = ProgressNotificationState.Completed;
                 return true;
-            }, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken);
         }
 
         /// <summary>

```

File: osu.Game/Database/LegacyModelExporter.cs [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+            {
+                realmAccess.Run(r =>
+                {
+                    TModel refetchModel = r.Find<TModel>(id);
+                    ExportToStream(refetchModel, stream, notify, cancellationToken);
+                });
+            }, cancellationToken).ContinueWith(t =>
+            {
+                if (t.IsFaulted)
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:141]: No logic to handle cancellation?

File: osu.Game/Database/LegacyModelExporter.cs [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+
+                notify.CompletionText = "Export Complete, Click to open the folder";
+                notify.State = ProgressNotificationState.Completed;
+                return true;
+            }, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Exports an item to Stream.
+        /// Override if custom export method is required.
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:156]: Doesn't make sense.

File: osu.Game/Database/LegacyModelExporter.cs [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+                    .GetFiles(string.Empty, $"{itemFilename}*{FileExtension}")
+                    .Concat(exportStorage.GetDirectories(string.Empty));
+            string filename = NamingUtils.GetNextBestFilename(existingExports, $"{itemFilename}{FileExtension}");
+            bool success = false;
+
+            ProgressNotification notification = new ProgressNotification
+            {
+                State = ProgressNotificationState.Active,
+                Text = "Exporting...",
+                CompletionText = "Export completed"
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:91]: Just remove this?

File: osu.Game/Database/LegacyModelExporter.cs [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+            // check if the model is being exporting already
+            if (!exportingModels.Contains(model))
+            {
+                exportingModels.Add(model);
+            }
+            else
+            {
+                PostNotification?.Invoke(new SimpleErrorNotification
+                {
+                    Text = "File is being exported"
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:74]: ???

Comment (peppy) [reply osu.Game/Database/LegacyModelExporter.cs:74]: Don't do this. Remove this weird handling.

File: osu.Game/Database/LegacyModelExporter.cs [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+            }, cancellationToken).ContinueWith(t =>
+            {
+                if (t.IsFaulted)
+                {
+                    notify.State = ProgressNotificationState.Cancelled;
+                    Logger.Error(t.Exception, "An error occurred while exporting");
+                    return false;
+                }
+
+                notify.CompletionText = "Export Complete, Click to open the folder";
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs:148]: Incorrect grammar. Also it doesn't really open the folder but present the file.

Maybe look at https://github.com/ppy/osu/blob/f8537c1cbe98bc38f0423e8eabdc2332a21d22ab/osu.Game/Database/RealmArchiveModelImporter.cs#L157-L172 for inspiration.

**Review by peppy — CHANGES_REQUESTED** [review]
zzzz

File: osu.Game/Database/LegacyModelExporter.cs (hunk 14 above)
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs:113]: If it is not finally, the exported error (An error other than `OperationCanceledException` has occurred) file should not be deleted.
If it is determined that keeping the wrong file is the expected behavior, then I will change it,

File: osu.Game/Database/LegacyModelExporter.cs (hunk 17 above)
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs:141]: because of `cancellationToken`, this task will be canceled
I have stayed this before:
```C#
if (t.IsCanceled)
   return;
```

File: osu.Game/Database/LegacyModelExporter.cs (hunk 18 above)
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs:156]: The previous compression logic is here (virtual method), forgot to change the comment

File: osu.Game/Database/LegacyModelExporter.cs (hunk 19 above)
Comment (cdwcgt) [reply osu.Game/Database/LegacyModelExporter.cs:91]: For `CompletionText`, I know it will be rewritten later, but I keep a generic text for other cases.
For example, the target platform does not support clicking to open the folder.

Maybe I'm overthinking it, but let me know if it's not necessary and I'll change it

File: osu.Game/Database/LegacyModelExporter.cs [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+using osu.Game.Overlays.Notifications;
+using osu.Game.Utils;
+using Realms;
+
+namespace osu.Game.Database
+{
+    /// <summary>
+    /// A class which handles exporting legacy user data of a single type from osu-stable.
+    /// </summary>
+    public abstract class LegacyModelExporter<TModel>
```
Comment (peppy) [inline osu.Game/Database/LegacyModelExporter.cs]: I'm renaming this back because the way you made changes and committed them makes it impossible to know what (if anything )in this class was changed.

**Comment by peppy** [comment]
I've basically rewritten this PR.

@bdach requesting your fresh eyes on it.
@cdwcgt please, read through each and every one of my commits multiple times and understand why the changes I made were made. ask if you have any questions.

I've left the rename in so the diff is legible (https://github.com/ppy/osu/pull/21308/commits/01581024678c6f4d99a54892cb95cbe7599f0f3c). It can be renamed back before merge.

**Review by peppy — DISMISSED** [review]

File: osu.Game/Database/LegacyExporter.cs [hunk 23]
```diff
@@ -38,56 +39,107 @@ public abstract class LegacyExporter<TModel>
...
         private readonly Storage exportStorage;
+        protected virtual string GetFilename(TModel item) => item.GetDisplayString();
+
+        public Action<Notification>? PostNotification { get; set; }
 
+        /// <summary>
+        /// Construct exporter.
+        /// Create a new exporter for each export, otherwise it will cause confusing notifications.
+        /// </summary>
+        /// <param name="storage">Storage for storing exported files. Basically it is used to provide export stream</param>
```
Comment (peppy) [inline osu.Game/Database/LegacyExporter.cs:52]: xmldoc still needs rewriting. i'll get to it later if i'm not beaten.

File: osu.Game/Database/LegacyExporter.cs [hunk 24]
```diff
@@ -38,56 +39,107 @@ public abstract class LegacyExporter<TModel>
...
+            Task.Run(() => { model.PerformRead(s => ExportToStream(s, stream, notification, cancellationToken)); }, cancellationToken);
+
+        /// <summary>
+        /// Exports model to Stream.
+        /// </summary>
+        /// <param name="model">The model to export.</param>
+        /// <param name="outputStream">The output stream to export to.</param>
+        /// <param name="notification">The notification will displayed to the user</param>
+        /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
+        protected abstract void ExportToStream(TModel model, Stream outputStream, ProgressNotification? notification, CancellationToken cancellationToken = default);
```
Comment (peppy) [inline osu.Game/Database/LegacyExporter.cs:143]: Probably no reason to be `protected`.

**Comment by cdwcgt** [comment]
https://github.com/ppy/osu/pull/21308/commits/2a3e03695cad820002999e420797de7a0faf4cba#diff-d5a620a82a32a4027cf0f259048441c48a70c4825f93636aa06a4c4aefb102caL48-L50
Shouldn't we prevent duplicate exports, like clicking twice "export the current skin"

**Comment by peppy** [comment]
I don't see any reason to prevent it.

**Comment by cdwcgt** [comment]
ok then I should be fine.
I will change `ExportToStream` change to public, and remove [this comment](https://github.com/ppy/osu/pull/21308/files#diff-79d1a2d699ac04e9d6b897b9116fbae2c542fadb0b1ba6c86f58995d467d15ceR72)

it should be ok after changing the class name, if no other problem.

File: osu.Game/Database/LegacyScoreExporter.cs [hunk 25]
```diff
@@ -28,7 +26,9 @@ protected override string GetFilename(ScoreInfo score)
             return filename;
         }
 
-        public override void ExportModelTo(ScoreInfo model, Stream outputStream)
+        protected override string FileExtension => ".osr";
+
+        public override void ExportToStream(ScoreInfo model, Stream outputStream, ProgressNotification? notification, CancellationToken cancellationToken = default)
```
Comment (bdach) [inline osu.Game/Database/LegacyScoreExporter.cs:31]: This isn't handling the case of missing file in the same way as `LegacyArchiveExporter` - but I'm looking away not to add more stuff onto of this trash fire of a branch than absolutely necessary (and I still want to push in a few *other* touch-ups).

**Review by bdach — APPROVED** [review]
